Chapters 1 & 2 Revisions 05-02-2025 (#5113)
dfinity/motoko: doc/md/fundamentals/2-types/6-objects-classes.md

## Objects

> In Motoko, an object is just a collection of named fields, holding values. These values can either be plain data, or function values. In addition, each field can be mutable or immutable.

Reviewer: ```suggestion
In Motoko, an object is a collection of named fields that hold values.  
These values can be plain data or functions.  
Each field can be either **mutable** or **immutable**, depending on whether it's declared with `var` or not.
```

A simple object containing just fields of data is like a record in a database.  

> Motoko's light-weight  [record](https://internetcomputer.org/docs/motoko/fundamentals/types/records) syntax make it easy do construct such objects..  

Reviewer: ```suggestion
Motoko's light-weight [record](https://internetcomputer.org/docs/motoko/fundamentals/types/records) syntax makes it easy to construct such objects.
```

When fields contain function values, Motoko objects can represent traditional objects with methods, familiar from object-oriented programming (OOP).  
From an OOP perspective, an object is an abstraction, defined by the behavior of its methods. Methods are typically used to modify or observe some encapsulated (i.e. hidden) state of an object.  

In addition to the record syntax, Motoko let's you defined objects from a block of declarations.  
The declarations in the block can be `public` or `private`, with `private` the default.  

> The public declarations become fields of the object, all private declarations are hidden.

Reviewer: ```suggestion
Public declarations become accessible fields of the object, while private declarations remain hidden and inaccessible from outside the object.
```

> ## Objects

Reviewer 2: All the examples are no-repl. Make em executable.

## Objects

> In Motoko, an object is a collection of named fields that hold values.  
> These values can be plain data or functions.  Each field can be either **mutable** or **immutable**, depending on whether it's declared with `var` or not.

Reviewer: ```suggestion
In Motoko, an object is a collection of named fields that hold values. These values can be plain data or functions.  Each field can be either **mutable** or **immutable** depending on whether it's declared with `var` or not.
```

> ## Objects

Author: Will have to discuss the exact details of this later.

The declarations in the block can be `public` or `private`, with `private` the default.  
The public declarations become fields of the object, all private declarations are hidden.  

> An object, similar to a [record](https://internetcomputer.org/docs/motoko/fundamentals/types/records), stores structured data with optional mutable fields and supports methods, including [asynchronous](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await) behavior. Unlike records, objects can encapsulate or share their [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and behavior using `public` and `private` visibility modifiers. However, they cannot be instantiated independently, such as `object()` use `object class()` instead.

Reviewer: ```suggestion
An object, similar to a [record](https://internetcomputer.org/docs/motoko/fundamentals/types/records), stores structured data with optional mutable fields and supports methods, including [asynchronous](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await) behavior. Unlike records, objects can encapsulate or share their [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and behavior using `public` and `private` visibility modifiers. However, they cannot be instantiated independently as `object()`.  Instead, you must use an object class (`object class()`).
```

An object class defines a blueprint for multiple objects. The above is just short-hand for an `object` class. Motoko also support module and actor classes.  

```motoko
object class Account(initialBalance : Nat) {
  var balance = initialBalance;

  public func deposit(amount : Nat) : Nat {
    balance += amount;
    balance
  };

  public func withdraw(amount : Nat) : ?Nat {
    if (amount > balance) { return null };
      balance -= amount;
      ?balance
    };
}
```

## Modules

Modules are similar to objects, containing public and private declarations, but are restricted to be stateless. They are typically used to implement libraries of types, functions and values, and, unlike objects, can be imported from other files.

```motoko name=float
import Float "mo:base/Float";

module CurrencyConverter {
  public func toUSD(amount : Nat) : Float {
    Float.fromInt(amount) * 1.1;
    };
};
```

### Module classes

A module class can be used to produce multiple modules with different configurations.

> ```motoko include=float

Reviewer: This does not work in the preview and returns an error

> ```motoko include=float

Reviewer: ```suggestion
```motoko no-repl
```

A function expecting `{ getBalance : () -> Nat }` can accept any of the above, since all contain at least that method. However, a function requiring `{ withdraw : Nat -> Bool }` cannot accept more general types that lack this method.

- In general, if type `T` is a subtype of type `U`, then any value of type `T` also has type `U`. The means that a value of the more specific type `T` can pass as a value of the more general type `U` without any explicit conversion.
- When `T` and `U` are object types, subtyping ensures that `T` provides at least the fields required by `U`, so that any object of type `T` can serve as an object of the more general type `U`.

The more general object type has fewer fields because this places fewer requirements on its values.

`PremiumAccount <: StandardAccount <: BasicAccount`, or equivalently:  `PremiumAccount` is a subtype of `StandardAccount`, which is a subtype of `BasicAccount`.

A function expecting `PremiumAccount` expects `withdraw`, so it cannot accept `basicAccount`.
However, a function expecting `BasicAccount` only needs `getBalance`, so it can accept all three type of objects.

> ```motoko include=accounts

Reviewer: This does not work in the preview and returns an error

> ```motoko include=accounts

Reviewer: ```suggestion
```motoko no-repl
```

import Debug "mo:base/Debug";
import Nat "mo:base/Nat";

func printBalance(account : { getBalance : () -> Nat }) {
  Debug.print("Balance: " # Nat.toText(account.getBalance()));
};

// Works, because all have getBalance
printBalance(_basicAccount);
printBalance(_standardAccount);
printBalance(_premiumAccount);
```

> ```motoko include=accounts

Reviewer: This does not work in the preview and returns an error

> ```motoko include=accounts

Reviewer: ```suggestion
```motoko no-repl
```

Debug.print(Float.toText(eurConverter.convert(100)));  // "90.0"
```

## Object subtyping

Object subtyping allows objects with more fields to be treated as subtypes of objects with fewer fields. This enables flexibility in function arguments and object compatibility.

Objects with fewer fields are more general, while objects with additional fields are subtypes of more general types.

> | Specificity Level | Description | Relationship |

Reviewer: ```suggestion
| Specificity level | Description | Relationship |
```

> - A subtype must be usable wherever its supertype is expected.
> - The more general object has fewer methods because it makes fewer assumptions about available functionality.

Reviewer 2: ```suggestion
- In general, if type `T` is a subtype of type `U`, then any value of type `T` also has type `U`. The means that a value of the more specific type `T` can pass as a value of the more general type `U` without any explicit conversion.
- When `T` and `U` are object types, subtyping ensures that `T` provides at least the fields required by `U`, so that any object of type `T` can serve as an object of the more general type `U`. 

The more general object type has fewer fields because this places fewer requirements on its values.
```

This is hard to explain. I hope that improves it a bit.
